include/mqtt/broker/security.hpp

```diff
@@ -0,0 +1,583 @@
+        std::set<std::string> username_and_groups;
+        username_and_groups.insert(std::string(username));
+
+        for (auto const &i: groups_) {
+            if (std::find(i.second.members.begin(), i.second.members.end(), username) != i.second.members.end()) {
+                username_and_groups.insert(i.first);
+            }
+        }
+
+        for (auto const &i: authorization_) {
+
+            if (i.sub_type != authorization::type::none) {
+                bool sets_intersect = false;
+                auto store_intersect = [&sets_intersect](std::string const &) mutable { sets_intersect = true; };
+
+                std::set_intersection(
+                    i.sub.begin(), i.sub.end(),
+                    username_and_groups.begin(), username_and_groups.end(),
+                    boost::make_function_output_iterator(std::ref(store_intersect))
+                );
+
+                if (sets_intersect) {
+                    callback(i);
+                }
+            }
+        }
+    }
+
+    authorization::type auth_pub(string_view const& topic, string_view const& username) const {
+        authorization::type result_type = authorization::type::deny;
+
+        auth_pub_map.find(topic, [&](std::string const &allowed_username, authorization::type type) {
+            if (allowed_username == username) result_type = type;
+        });
+
+        return result_type;
+    }
+
+    std::map<std::string, authorization::type> auth_sub(string_view const& topic) const {
+        std::map<std::string, authorization::type> result;
+
+        auth_sub_map.find(topic, [&](std::string const &allowed_username, authorization::type type) {
+            result[allowed_username] = type;
+        });
+
+        return result;
+    }
+
+    static authorization::type auth_sub_user(std::map<std::string, authorization::type> const& result, std::string const& username) {
+        auto i = result.find(username);
+        if (i == result.end()) return authorization::type::deny;
+        return i->second;
+    }
+
+    static bool is_hash(string_view const &level) { return level == "#"; }
+    static bool is_plus(string_view const &level) { return level == "+"; }
+    static bool is_literal(string_view const &level) { return !is_hash(level) && !is_plus(level); }
+
+    static optional<std::string> is_subscribe_allowed(std::vector<std::string> const &authorized_filter, string_view const &subscription_filter) {
+        optional<std::string> result;
+        auto append_result = [&result](string_view const &token) {
+              if (result) {
+                  result.value() += topic_filter_separator;
+                  result.value().append(token.data(), token.size());
+              } else {
+                  result = optional<std::string>(token);
+              }
+          };
+
+        auto filter_begin = authorized_filter.begin();
+
+        auto subscription_begin = std::prev(subscription_filter.begin());
```

Comment:
prev of begin is not good even if the iterator isn't dereferenced. It is slightly different from end() (past the last element).
Getting past the last iterator by end() is safe, but dereference is UB. On the other hand, getting the prev of begin() iterator itself is UB.
See https://stackoverflow.com/questions/25263676/is-it-safe-to-use-stdprevvector-begin-or-stdnextvector-begin-1-lik

Reply:
I have not been able to successfully fix this yet

Reply:
Maybe the algorithm need to modified. I guess that the code would become a little less elegant but it is needed.

Reply:
i resolved the issue, it is still an elegant approach, i would say.